pkg/xrd/gen.go

```diff
@@ -0,0 +1,318 @@
+
+	if len(xrdVersions) == 0 {
+		xrdVersions = []string{defaultVersion}
+	}
+
+	var headerText string
+
+	if g.HeaderFile != "" {
+		headerBytes, err := ctx.ReadFile(g.HeaderFile)
+		if err != nil {
+			return err
+		}
+		headerText = string(headerBytes)
+	}
+	headerText = strings.ReplaceAll(headerText, " YEAR", " "+g.Year)
+
+	for _, groupKind := range kubeKinds {
+		parser.NeedXRDFor(groupKind, g.MaxDescLen)
+		xrdRaw := parser.XRDefinitons[groupKind]
+
+		//addAttribution(&crdRaw)
+
+		// the XRD status should be embedded as a field but we can suppress it in the XRD generation
+		// and let crossplane inject it for us
+		for _, version := range xrdRaw.Spec.Versions {
+			removeXRDStatusProps(version.Schema.OpenAPIV3Schema)
+		}
+
+		fileName := fmt.Sprintf("%s_%s.yaml", xrdRaw.Spec.Group, xrdRaw.Spec.Names.Plural)
+		if err := ctx.WriteYAML(fileName, headerText, []interface{}{xrdRaw}, genall.WithTransform(transformRemoveXRDStatus)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func removeDescriptionFromMetadata(crd *apiext.CustomResourceDefinition) {
+	for _, versionSpec := range crd.Spec.Versions {
+		if versionSpec.Schema != nil {
+			removeDescriptionFromMetadataProps(versionSpec.Schema.OpenAPIV3Schema)
+		}
+	}
+}
+
+func removeDescriptionFromMetadataProps(v *apiext.JSONSchemaProps) {
+	if m, ok := v.Properties["metadata"]; ok {
+		meta := &m
+		if meta.Description != "" {
+			meta.Description = ""
+			v.Properties["metadata"] = m
+
+		}
+	}
+}
+
+// FixTopLevelMetadata resets the schema for the top-level metadata field which is needed for CRD validation
+func FixTopLevelMetadata(crd apiext.CustomResourceDefinition) {
+	for _, v := range crd.Spec.Versions {
+		if v.Schema != nil && v.Schema.OpenAPIV3Schema != nil && v.Schema.OpenAPIV3Schema.Properties != nil {
+			schemaProperties := v.Schema.OpenAPIV3Schema.Properties
+			if _, ok := schemaProperties["metadata"]; ok {
+				schemaProperties["metadata"] = apiext.JSONSchemaProps{Type: "object"}
+			}
+		}
+	}
+}
+
+// addAttribution adds attribution info to indicate controller-gen tool was used
+// to generate this CRD definition along with the version info.
+func addAttribution(crd *apiext.CustomResourceDefinition) {
+	if crd.ObjectMeta.Annotations == nil {
+		crd.ObjectMeta.Annotations = map[string]string{}
+	}
+	crd.ObjectMeta.Annotations["controller-gen.kubebuilder.io/version"] = version.Version()
+}
+
+// FindMetav1 locates the actual package representing metav1 amongst
+// the imports of the roots.
+func FindMetav1(roots []*loader.Package) *loader.Package {
+	for _, root := range roots {
+		pkg := root.Imports()["k8s.io/apimachinery/pkg/apis/meta/v1"]
+		if pkg != nil {
+			return pkg
+		}
+	}
+	return nil
+}
+
+// FindKubeKinds locates all types that contain TypeMeta and ObjectMeta
+// (and thus may be a Kubernetes object), and returns the corresponding
+// group-kinds.
+func FindKubeKinds(parser *crd.Parser, metav1Pkg *loader.Package) []schema.GroupKind {
```

Comment:
This seems like the same function from `pkg/crds`?  Could this function simply be referenced via import?

Reply:
Probably. Most of this file was cribbed from pkg/crd/gen.go but there is a mix of public and private things in crd/gen and crd/parser. 

This function might need to change for pkg/xrd to split between XRDs and Claim types. As is I can move it back to use the imported version.

Reply:
Makes sense. I was wondering if I was missing something since they looked the same to me.  I agree that at some point these may differ.